packages/libs/deer/desert/src/token.rs

```diff
@@ -3,69 +3,202 @@ use core::fmt::{Debug, Display, Formatter};
-    /// struct Bool(bool);
-    ///
-    /// impl Reflection for Bool {
-    ///     fn schema(_: &mut Document) -> Schema {
-    ///         Schema::new("boolean")
-    ///     }
+    /// #[derive(Debug, PartialEq, serde::Serialize)]
+    /// pub enum LogLevel {
+    ///     Trace,
+    ///     Debug,
+    ///     Info,
+    ///     Warn,
+    ///     Error,
+    ///     Critical,
+    ///     Unknown,
     /// }
     ///
-    /// impl<'de> Deserialize<'de> for Bool {
-    ///     type Reflection = Self;
+    /// struct LogLevelVisitor;
     ///
-    ///     fn deserialize<D: Deserializer<'de>>(de: D) -> error_stack::Result<Self, DeserializeError> {
-    ///         struct BoolVisitor;
+    /// impl<'de> Visitor<'de> for LogLevelVisitor {
+    ///     type Value = LogLevel;
     ///
-    ///         impl<'de> Visitor<'de> for BoolVisitor {
-    ///             type Value = Bool;
-    ///
-    ///             fn expecting(&self) -> Document {
-    ///                 Bool::reflection()
-    ///             }
+    ///     fn expecting(&self) -> Document {
+    ///         Self::Value::reflection()
+    ///     }
     ///
-    ///             fn visit_bool(self, v: bool) -> error_stack::Result<Self::Value, VisitorError> {
-    ///                 Ok(Bool(v))
-    ///             }
+    ///     fn visit_str(self, v: &str) -> error_stack::Result<Self::Value, VisitorError> {
+    ///         match v {
+    ///             "trace" => Ok(LogLevel::Trace),
+    ///             "debug" => Ok(LogLevel::Debug),
+    ///             "info" => Ok(LogLevel::Info),
+    ///             "warn" => Ok(LogLevel::Warn),
+    ///             "error" => Ok(LogLevel::Error),
+    ///             "critical" => Ok(LogLevel::Critical),
+    ///             _ => Ok(LogLevel::Unknown),
     ///         }
+    ///     }
+    /// }
+    ///
+    /// impl Reflection for LogLevel {
+    ///     fn schema(doc: &mut Document) -> Schema {
+    ///         Schema::new("string").with("enum", [
+    ///             "trace", "debug", "info", "warn", "error", "critical",
+    ///         ])
+    ///     }
+    /// }
     ///
-    ///         de.deserialize_bool(BoolVisitor)
+    /// impl<'de> Deserialize<'de> for LogLevel {
+    ///     type Reflection = Self;
+    ///
+    ///     fn deserialize<D: Deserializer<'de>>(de: D) -> error_stack::Result<Self, DeserializeError> {
+    ///         de.deserialize_str(LogLevelVisitor)
     ///             .change_context(DeserializeError)
     ///     }
     /// }
     ///
-    /// assert_tokens(&Bool(true), &[Token::Bool(true)])
+    /// assert_tokens(&LogLevel::Critical, &[Token::Str("critical")])
     /// ```
-    Bool(bool),
-    Number(&'static Number),
-    Char(char),
     Str(&'static str),
+
+    /// A serialized [`str`]
+    ///
+    /// ```
+    /// use deer_desert::{assert_tokens, Token};
+    /// let test = "example";
+    ///
+    /// assert_tokens(&test, &[Token::BorrowedStr(test)])
+    /// ```
     BorrowedStr(&'static str),
```

Comment:
Would the `Str` variant have the same behavior as this one? Not sure what the difference between the two would be in practise

Reply:
No, I tried to visualise this with the example implementation of `Str`, there are deserializers out there that might not be able to give straight references to their input value. Therefore we cannot guarantee that every string will be borrowed from the input. Some deserializers might be consuming their input or are doing some hocus-pocus and therefore aren't able to guarantee `'de` as a lifetime.

TL;DR:

```
struct X<'de> { x: &'de str }
```

Only works if the deserializer supports borrowed strings, while:

```
struct Y { x: A };
/* deserializer of A implement visit_str instead */
```

Works for both borrowed and regular strings.